PR: https://github.com/odoo/odoo/pull/233963

From: a9b8b2009eaf4299584501b2d9b91621f562f844
From: Aurélien Warnon
Date: 2026-02-14 01:01:27

Breaking data model changes scores: del:1 + add:0, change matches:
-    _inherit = 'utm.medium'

Total Changes: 26

[IMP] utm: improve and generalize UTM trackers data

RATIONALE

The way we currently use the UTM trackers (aka source/medium/campaign) does not
comply with the industry standards, leading to confusion and globally unclear
or not usable reporting on UTM data.

PURPOSE

Clear the confusion of having specific social media as UTM medium.
The UTM medium should always be "Social Media" (general communication channel).

In addition, we also add the specific social media as UTM sources, for global
usage by other apps (notably Social).

To protect those common records and to prevent them to be deleted, we extend
the previous system to work for all utm models (source and medium notably).
Including making it less verbose and easier to use with "_utm_ref".

CLEANING

We also used this opportunity to clean unused or irrelevant UTM data records.
For example, "Banner" and "Television" were globally unused and do not make a
lot of sense anymore as data.

As a result, demo lead records were updated to use different UTM.

(Note: some of the demo do not make a lot of sense, such as phone + Google, but
the scope of this PR is already big enough and it can be reworked later).

Task-3330036

Part-of: odoo/odoo#233963
Related: odoo/enterprise#98569
Related: odoo/upgrade#8746
Signed-off-by: Thibault Delavallee (tde) <tde@openerp.com>

================================= pseudo patch: =================================

--- a/addons/mass_mailing_sms/models/mailing_mailing.py
+++ b/addons/mass_mailing_sms/models/mailing_mailing.py
@@ -61,10 +61,10 @@ class MailingMailing(models.Model):
     def _compute_medium_id(self):
         super()._compute_medium_id()
         for mailing in self:
-            if mailing.mailing_type == 'sms' and (not mailing.medium_id or mailing.medium_id == self.env['utm.medium']._fetch_or_create_utm_medium('email')):
-                mailing.medium_id = self.env['utm.medium']._fetch_or_create_utm_medium("sms", module="mass_mailing_sms").id
-            elif mailing.mailing_type == 'mail' and (not mailing.medium_id or mailing.medium_id == self.env['utm.medium']._fetch_or_create_utm_medium("sms", module="mass_mailing_sms")):
-                mailing.medium_id = self.env['utm.medium']._fetch_or_create_utm_medium('email').id
+            if mailing.mailing_type == 'sms' and (not mailing.medium_id or mailing.medium_id == self.env['utm.mixin']._utm_ref('utm.utm_medium_email')):
+                mailing.medium_id = self.env['utm.mixin']._utm_ref('utm.utm_medium_sms').id
+            elif mailing.mailing_type == 'mail' and (not mailing.medium_id or mailing.medium_id == self.env['utm.mixin']._utm_ref('utm.utm_medium_sms')):
+                mailing.medium_id = self.env['utm.mixin']._utm_ref('utm.utm_medium_email').id
 
     @api.depends('sms_template_id', 'mailing_type')
     def _compute_body_plaintext(self):

--- a/addons/mass_mailing_sms/models/utm.py
+++ b/addons/mass_mailing_sms/models/utm.py
@@ -77,19 +77,9 @@ class UtmCampaign(models.Model):
         return ab_testing_campaign
 
 
-class UtmMedium(models.Model):
-    _inherit = 'utm.medium'
-
-    @api.ondelete(at_uninstall=False)
-    def _unlink_except_utm_medium_sms(self):
-        utm_medium_sms = self.env.ref('mass_mailing_sms.utm_medium_sms', raise_if_not_found=False)
-        if utm_medium_sms and utm_medium_sms in self:
-            raise UserError(_(
-                "The UTM medium '%s' cannot be deleted as it is used in some main "
-                "functional flows, such as the SMS Marketing.",
-                utm_medium_sms.name
-            ))
+class UtmMixin(models.Model):
+    _inherit = ['utm.mixin']
 
     @property
-    def SELF_REQUIRED_UTM_MEDIUMS_REF(self):
-        return super().SELF_REQUIRED_UTM_MEDIUMS_REF | {"mass_mailing_sms.utm_medium_sms": "SMS"}
+    def SELF_REQUIRED_UTM_REF(self):
+        return super().SELF_REQUIRED_UTM_REF | {'mass_mailing_sms.utm_medium_sms': ('SMS', 'utm.medium')}
